x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/MlPlatformArchitecturesUtil.java

```diff
@@ -0,0 +1,127 @@
+
+import static org.elasticsearch.core.Strings.format;
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+public class MlPlatformArchitecturesUtil {
+
+    public static void getMlNodesArchitecturesSet(ActionListener<Set<String>> architecturesListener, Client client, ThreadPool threadPool) {
+        ActionListener<NodesInfoResponse> listener = MlPlatformArchitecturesUtil.getArchitecturesSetFromNodesInfoResponseListener(
+            threadPool,
+            architecturesListener
+        );
+
+        NodesInfoRequest request = MlPlatformArchitecturesUtil.getNodesInfoBuilderWithMlNodeArchitectureInfo(client).request();
+        executeAsyncWithOrigin(client, ML_ORIGIN, NodesInfoAction.INSTANCE, request, listener);
+    }
+
+    static ActionListener<NodesInfoResponse> getArchitecturesSetFromNodesInfoResponseListener(
+        ThreadPool threadPool,
+        ActionListener<Set<String>> architecturesListener
+    ) {
+        return ActionListener.wrap(nodesInfoResponse -> {
+            threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
+                architecturesListener.onResponse(getArchitecturesSetFromNodesInfoResponse(nodesInfoResponse));
+            });
+        }, architecturesListener::onFailure);
+    }
+
+    static NodesInfoRequestBuilder getNodesInfoBuilderWithMlNodeArchitectureInfo(Client client) {
+        return client.admin().cluster().prepareNodesInfo().clear().setNodesIds("ml:true").setOs(true).setPlugins(true);
+    }
+
+    private static Set<String> getArchitecturesSetFromNodesInfoResponse(NodesInfoResponse nodesInfoResponse) {
+        return nodesInfoResponse.getNodes()
+            .stream()
+            .filter(node -> node.getNode().hasRole(DiscoveryNodeRole.ML_ROLE.roleName()))
+            .map(node -> {
+                OsInfo osInfo = node.getInfo(OsInfo.class);
+                return Platforms.platformName(osInfo.getName(), osInfo.getArch());
+            })
+            .collect(Collectors.toUnmodifiableSet());
+    }
+
+    public static void verifyMlNodesAndModelArchitectures(
+        ActionListener<Void> failureListener,
+        Client client,
+        ThreadPool threadPool,
+        String modelPlatformArchitecture,
+        String modelID
+    ) {
+        String modifiedPlatformArchitecture = (modelPlatformArchitecture == null && modelID.contains("linux-x86_64"))
+            ? "linux-x86_64"
+            : null;
+        ActionListener<Set<String>> architecturesListener = ActionListener.wrap((architectures) -> {
+            verifyMlNodesAndModelArchitectures(architectures, modifiedPlatformArchitecture, modelID);
+        }, failureListener::onFailure);
+
+        getMlNodesArchitecturesSet(architecturesListener, client, threadPool);
+    }
+
+    static void verifyMlNodesAndModelArchitectures(Set<String> architectures, String modelPlatformArchitecture, String modelID)
```

Comment:
nit: We might be able to structure this with less if-block nesting

Maybe something like this?

```
        Iterator<String> architecturesIterator = architectures.iterator();

        if (modelPlatformArchitecture == null || architectures.isEmpty() || architecturesIterator.hasNext() == false) {
            return;
        }

        if (architectures.size() > 1) {
            throw new IllegalStateException(
                format(
                    "ML nodes in this cluster have multiple platform architectures, but can only have one for this model ([%s]); "
                        + "expected [%s]; "
                        + "but was %s",
                    modelID,
                    modelPlatformArchitecture,
                    architectures
                )
            );
        }

        if (Objects.equals(architecturesIterator.next(), modelPlatformArchitecture) == false) {

            throw new IllegalArgumentException(
                format(
                    "The model being deployed ([%s]) is platform specific and incompatible with ML nodes in the cluster; "
                        + "expected [%s]; "
                        + "but was %s",
                    modelID,
                    modelPlatformArchitecture,
                    architectures
                )
            );
        }
```

Reply:
We definitely could do it that way. I personally prefer not to mix return statements with if statements when possible (i.e., I prefer if block nesting to returning early), as I find it easy to miss a single return statement, but if you think its better that way I am happy to make this change.

Reply:
Fair point, you can leave it as is. In Kibana the style guide actually called out preferring early return: https://github.com/elastic/kibana/blob/main/STYLEGUIDE.mdx#L235

I didn't see anything in elasticsearch that states it though.

Reply:
Thanks for sharing. Looking at this some more, I think it would be good to return early in this case, since each code-block does return or throw, I think your version is clearer. I will go with your suggestion.